yarn-project/noir-protocol-circuits/src/crates/types/src/header.nr

```diff
@@ -1,15 +1,98 @@
 use crate::{
     abis::{
-        append_only_tree_snapshot::AppendOnlyTreeSnapshot,
-        global_variables::GlobalVariables,
+        append_only_tree_snapshot::{
+            AppendOnlyTreeSnapshot,
+            APPEND_ONLY_TREE_SNAPSHOT_LENGTH,
+        },
+        global_variables::{
+            GlobalVariables,
+            GLOBAL_VARIABLES_LENGTH,
+        },
+    },
+    constants::{
+        GENERATOR_INDEX__BLOCK_HASH,
+        HEADER_LENGTH,
+        NUM_FIELDS_PER_SHA256,
+    },
+    hash::pedersen_hash,
+    state_reference::{
+        StateReference,
+        STATE_REFERENCE_LENGTH,
+    },
+    traits::Empty,
+    utils::{
+        arr_copy_slice,
+        bounded_vec::BoundedVec,
     },
-    constants::NUM_FIELDS_PER_SHA256,
-    state_reference::StateReference,
 };
 
+// docs:start:header
 struct Header {
     last_archive: AppendOnlyTreeSnapshot,
     body_hash: [Field; NUM_FIELDS_PER_SHA256],
     state: StateReference,
     global_variables: GlobalVariables,
 }
+// docs:end:header
+
+impl Header {
+
+    pub fn serialize(self) -> [Field; HEADER_LENGTH] {
+        let mut fields: BoundedVec<Field, HEADER_LENGTH> = BoundedVec::new(0);
+
+        fields.push_array(self.last_archive.serialize());
+        fields.push_array(self.body_hash);
+        fields.push_array(self.state.serialize());
+        fields.push_array(self.global_variables.serialize());
+
+        fields.storage
+    }
+
+    pub fn deserialize(serialized: [Field; HEADER_LENGTH]) -> Self {
+        let mut offset = 0;
+
+        let last_archive_fields = arr_copy_slice(serialized, [0; APPEND_ONLY_TREE_SNAPSHOT_LENGTH], offset);
+        offset = offset + APPEND_ONLY_TREE_SNAPSHOT_LENGTH;
+
+        let body_hash = arr_copy_slice(serialized, [0; NUM_FIELDS_PER_SHA256], offset);
+        offset = offset + NUM_FIELDS_PER_SHA256;
+
+        let state_fields = arr_copy_slice(serialized, [0; STATE_REFERENCE_LENGTH], offset);
+        offset = offset + STATE_REFERENCE_LENGTH;
+
+        let global_variables_fields = arr_copy_slice(serialized, [0; GLOBAL_VARIABLES_LENGTH], offset);
+
+        Header {
+            last_archive: AppendOnlyTreeSnapshot::deserialize(last_archive_fields),
+            body_hash,
+            state: StateReference::deserialize(state_fields),
+            global_variables: GlobalVariables::deserialize(global_variables_fields),
+        }
+    }
+
+    // TODO: Nuke this. This is not the desired block hash. Keeping it here for compatibility reasons.
+    pub fn block_hash(self) -> Field {
```

Comment:
What compatibility are we keeping this for?

Reply:
It's incorrectly formulated. Mostly meant that I don't want to update it in this PR as it would be out of scope. Updated the comment and linked the relevant issue in https://github.com/AztecProtocol/aztec-packages/pull/4154/commits/9a72c78c7983fea0b2c35aab8f6a412f3b9dbfb3